Some binaries require GOPATH mode to be installed

### What did you do? (required. The issue will be **closed** when not provided.)
:GoInstallBinaries

### What did you expect to happen?
Install vim-go properly

### What happened instead?
vim-go is not installed

### Configuration (**MUST** fill this out):

* vim-go version:
I cannot get this as vim-go doesn't install. I am up-to-date (as of 27/04/2019) in cloned git repo of vim-go

* `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
set smartindent
set tabstop=2
set shiftwidth=4
set expandtab
set nocompatible
set shell=/bin/sh
syntax on
filetype plugin on
let g:go_disable_autoinstall = 0
" Use a blinking upright bar cursor in Insert mode, a blinking block in normal
if &term == 'xterm-256color' || &term == 'screen-256color'
  let &t_SI = "\<Esc>[5 q"
  let &t_EI = "\<Esc>[1 q"
endif

if exists('$TMUX')
  let &t_EI = "\<Esc>Ptmux;\<Esc>\<Esc>]50;CursorShape=0\x7\<Esc>\\"
  let &t_SI = "\<Esc>Ptmux;\<Esc>\<Esc>]50;CursorShape=1\x7\<Esc>\\"
endif

* Vim version (first three lines from `:version`):
VIM - Vi IMproved 8.1 (2018 May 18, compiled Feb 19 2019 12:07:03)
macOS version
Included patches: 1-950
Compiled by Homebrew

* Go version (`go version`):
go version go1.12 darwin/amd64

* Go environment (`go env`):
GOARCH="amd64"
GOBIN="/Users/Andrew/go/bin"
GOCACHE="/Users/Andrew/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/Andrew/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/opt/go/libexec"
GOTMPDIR=""
GOTOOLDIR="/usr/local/opt/go/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/2h/fs20m5d15p59qn_30g437b000000gn/T/go-build304997520=/tmp/go-build -gno-record-gcc-switches -fno-common"

This is the error I get about gometalinter

```vim-go: gometalinter not found. Installing github.com/alecthomas/gometalinter to folder /Users/Andrew/go/bin
Error installing github.com/alecthomas/gometalinter: go: finding github.com/google/shlex latest^@go: finding gopkg.in/alecthomas/kingpin.v3-unstable latest^@go: finding github.com/nicksnyder/go-i18n/i18n latest^@go: finding github.com/alecthomas/units latest^@# github.com
/alecthomas/gometalinter^@../pkg/mod/github.com/alecthomas/gometalinter@v3.0.0+incompatible/main.go:36:94: cannot use loadConfig (type func(*kingpin.Application, *kingpin.ParseElement, *kingpin.ParseContext) error) as type kingpin.Action in argument to app.cmdMixin.flagGr
oup.Flag("config", "Load JSON configuration from file.").Envar("GOMETALINTER_CONFIG").Action^@../pkg/mod/github.com/alecthomas/gometalinter@v3.0.0+incompatible/main.go:38:100: cannot use disableAction (type func(*kingpin.Application, *kingpin.ParseElement, *kingpin.ParseC
ontext) error) as type kingpin.Action in argument to app.cmdMixin.flagGroup.Flag("disable", "Disable previously enabled linters.").PlaceHolder("LINTER").Short('D').Action^@../pkg/mod/github.com/alecthomas/gometalinter@v3.0.0+incompatible/main.go:39:99: cannot use enableAc
tion (type func(*kingpin.Application, *kingpin.ParseElement, *kingpin.ParseContext) error) as type kingpin.Action in argument to app.cmdMixin.flagGroup.Flag("enable", "Enable previously disabled linters.").PlaceHolder("LINTER").Short('E').Action^@../pkg/mod/github.com/ale
cthomas/gometalinter@v3.0.0+incompatible/main.go:40:83: cannot use cliLinterOverrides (type func(*kingpin.Application, *kingpin.ParseElement, *kingpin.ParseContext) error) as type kingpin.Action in argument to app.cmdMixin.flagGroup.Flag("linter", "Define a linter.").Plac
eHolder("NAME:COMMAND:PATTERN").Action^@../pkg/mod/github.com/alecthomas/gometalinter@v3.0.0+incompatible/main.go:43:56: cannot use disableAllAction (type func(*kingpin.Application, *kingpin.ParseElement, *kingpin.ParseContext) error) as type kingpin.Action in argument to
 app.cmdMixin.flagGroup.Flag("disable-all", "Disable all linters.").Action^@../pkg/mod/github.com/alecthomas/gometalinter@v3.0.0+incompatible/main.go:44:54: cannot use enableAllAction (type func(*kingpin.Application, *kingpin.ParseElement, *kingpin.ParseContext) error) as
 type kingpin.Action in argument to app.cmdMixin.flagGroup.Flag("enable-all", "Enable all linters.").Action^@../pkg/mod/github.com/alecthomas/gometalinter@v3.0.0+incompatible/main.go:200:12: cannot use loadDefaultConfig (type func(*kingpin.Application, *kingpin.ParseEleme
nt, *kingpin.ParseContext) error) as type kingpin.Action in argument to app.Action^@```

When I run :GoUpdateBinaries I get the aforementioned issue about gometalinter and the following issue about golangci-lint

```vim-go: Updating golangci-lint. Reinstalling github.com/golangci/golangci-lint/cmd/golangci-lint to folder /Users/Andrew/go/bin
Error downloading github.com/golangci/golangci-lint/cmd/golangci-lint: go: finding github.com/golangci/golangci-lint/cmd/golangci-lint latest^@go: finding github.com/golangci/golangci-lint/cmd latest^@Fetching https://gopkg.in/airbrake/gobrake.v2?go-get=1^@Fetching https:
//golang.org/x/tools?go-get=1^@Fetching https://golang.org/x/net?go-get=1^@go: finding github.com/golangci/ineffassign latest^@Fetching https://mvdan.cc/unparam?go-get=1^@go: finding github.com/golangci/go-tools latest^@Parsing meta tags from https://golang.org/x/tools?go
-get=1 (status code 200)^@get "golang.org/x/tools": found meta tag get.metaImport{Prefix:"golang.org/x/tools", VCS:"git", RepoRoot:"https://go.googlesource.com/tools"} at https://golang.org/x/tools?go-get=1^@Parsing meta tags from https://gopkg.in/airbrake/gobrake.v2?go-g
et=1 (status code 200)^@get "gopkg.in/airbrake/gobrake.v2": found meta tag get.metaImport{Prefix:"gopkg.in/airbrake/gobrake.v2", VCS:"git", RepoRoot:"https://gopkg.in/airbrake/gobrake.v2"} at https://gopkg.in/airbrake/gobrake.v2?go-get=1^@Parsing meta tags from https://go
lang.org/x/net?go-get=1 (status code 200)^@get "golang.org/x/net": found meta tag get.metaImport{Prefix:"golang.org/x/net", VCS:"git", RepoRoot:"https://go.googlesource.com/net"} at https://golang.org/x/net?go-get=1^@go: finding github.com/shirou/w32 latest^@go: finding g
ithub.com/mitchellh/go-ps latest^@Fetching https://sourcegraph.com/sourcegraph/go-diff?go-get=1^@go: finding github.com/logrusorgru/aurora latest^@go: finding github.com/golangci/check latest^@Fetching https://golang.org/x/text?go-get=1^@Parsing meta tags from https://gol
ang.org/x/text?go-get=1 (status code 200)^@get "golang.org/x/text": found meta tag get.metaImport{Prefix:"golang.org/x/text", VCS:"git", RepoRoot:"https://go.googlesource.com/text"} at https://golang.org/x/text?go-get=1^@Parsing meta tags from https://mvdan.cc/unparam?go-
get=1 (status code 200)^@get "mvdan.cc/unparam": found meta tag get.metaImport{Prefix:"mvdan.cc/unparam", VCS:"git", RepoRoot:"https://github.com/mvdan/unparam"} at https://mvdan.cc/unparam?go-get=1^@go: finding golang.org/x/tools latest^@Parsing meta tags from https://so
urcegraph.com/sourcegraph/go-diff?go-get=1 (status code 200)^@get "sourcegraph.com/sourcegraph/go-diff": found meta tag get.metaImport{Prefix:"sourcegraph.com/sourcegraph/go-diff", VCS:"git", RepoRoot:"https://github.com/sourcegraph/go-diff"} at https://sourcegraph.com/so
urcegraph/go-diff?go-get=1^@go: finding github.com/golangci/dupl latest^@go: finding github.com/golangci/errcheck latest^@go: finding golang.org/x/net latest^@Fetching https://sourcegraph.com/sqs/pbtypes?go-get=1^@Fetching https://golang.org/x/crypto?go-get=1^@Parsing met
a tags from https://sourcegraph.com/sqs/pbtypes?go-get=1 (status code 200)^@Parsing meta tags from https://golang.org/x/crypto?go-get=1 (status code 200)^@get "sourcegraph.com/sqs/pbtypes": found meta tag get.metaImport{Prefix:"sourcegraph.com/sqs/pbtypes", VCS:"git", Rep
oRoot:"https://github.com/sqs/pbtypes"} at https://sourcegraph.com/sqs/pbtypes?go-get=1^@go: finding mvdan.cc/unparam latest^@get "golang.org/x/crypto": found meta tag get.metaImport{Prefix:"golang.org/x/crypto", VCS:"git", RepoRoot:"https://go.googlesource.com/crypto"} a
t https://golang.org/x/crypto?go-get=1^@go: finding github.com/golangci/go-misc latest^@Fetching https://golang.org/x/sys?go-get=1^@Parsing meta tags from https://golang.org/x/sys?go-get=1 (status code 200)^@go: finding github.com/golangci/revgrep latest^@get "golang.org/
x/sys": found meta tag get.metaImport{Prefix:"golang.org/x/sys", VCS:"git", RepoRoot:"https://go.googlesource.com/sys"} at https://golang.org/x/sys?go-get=1^@go: finding github.com/golangci/gosec latest^@go: finding github.com/golangci/maligned latest^@Fetching https://go
pkg.in/check.v1?go-get=1^@go: finding github.com/mozilla/tls-observatory latest^@go: finding golang.org/x/crypto latest^@Fetching https://mvdan.cc/interfacer?go-get=1^@Fetching https://gopkg.in/gemnasium/logrus-airbrake-hook.v2?go-get=1^@Parsing meta tags from https://gop
kg.in/check.v1?go-get=1 (status code 200)^@get "gopkg.in/check.v1": found meta tag get.metaImport{Prefix:"gopkg.in/check.v1", VCS:"git", RepoRoot:"https://gopkg.in/check.v1"} at https://gopkg.in/check.v1?go-get=1^@Parsing meta tags from https://gopkg.in/gemnasium/logrus-a
irbrake-hook.v2?go-get=1 (status code 200)^@get "gopkg.in/gemnasium/logrus-airbrake-hook.v2": found meta tag get.metaImport{Prefix:"gopkg.in/gemnasium/logrus-airbrake-hook.v2", VCS:"git", RepoRoot:"https://gopkg.in/gemnasium/logrus-airbrake-hook.v2"} at https://gopkg.in/g
emnasium/logrus-airbrake-hook.v2?go-get=1^@Parsing meta tags from https://mvdan.cc/interfacer?go-get=1 (status code 200)^@get "mvdan.cc/interfacer": found meta tag get.metaImport{Prefix:"mvdan.cc/interfacer", VCS:"git", RepoRoot:"https://github.com/mvdan/interfacer"} at h
ttps://mvdan.cc/interfacer?go-get=1^@go: finding github.com/golangci/gocyclo latest^@go: finding golang.org/x/sys latest^@Fetching https://gopkg.in/yaml.v2?go-get=1^@go: finding github.com/golangci/gofmt latest^@Parsing meta tags from https://gopkg.in/yaml.v2?go-get=1 (st
atus code 200)^@get "gopkg.in/yaml.v2": found meta tag get.metaImport{Prefix:"gopkg.in/yaml.v2", VCS:"git", RepoRoot:"https://gopkg.in/yaml.v2"} at https://gopkg.in/yaml.v2?go-get=1^@go: finding mvdan.cc/interfacer latest^@go: finding gopkg.in/check.v1 latest^@go: finding
 github.com/shurcooL/go-goon latest^@go: finding github.com/golangci/unconvert latest^@Fetching https://gopkg.in/tomb.v1?go-get=1^@Parsing meta tags from https://gopkg.in/tomb.v1?go-get=1 (status code 200)^@go: finding github.com/shurcooL/go latest^@get "gopkg.in/tomb.v1"
: found meta tag get.metaImport{Prefix:"gopkg.in/tomb.v1", VCS:"git", RepoRoot:"https://gopkg.in/tomb.v1"} at https://gopkg.in/tomb.v1?go-get=1^@go: finding github.com/OpenPeeDeeP/depguard latest^@Fetching https://gopkg.in/errgo.v2?go-get=1^@Parsing meta tags from https:/
/gopkg.in/errgo.v2?go-get=1 (status code 200)^@get "gopkg.in/errgo.v2": found meta tag get.metaImport{Prefix:"gopkg.in/errgo.v2", VCS:"git", RepoRoot:"https://gopkg.in/errgo.v2"} at https://gopkg.in/errgo.v2?go-get=1^@go: finding gopkg.in/tomb.v1 latest^@go: finding githu
b.com/golangci/lint-1 latest^@go: finding github.com/StackExchange/wmi latest^@Fetching https://gopkg.in/fsnotify.v1?go-get=1^@go: sourcegraph.com/sourcegraph/go-diff@v0.5.1: parsing go.mod: unexpected module path "github.com/sourcegraph/go-diff"^@Parsing meta tags from h
ttps://gopkg.in/fsnotify.v1?go-get=1 (status code 200)^@go: finding github.com/golangci/goconst latest^@get "gopkg.in/fsnotify.v1": found meta tag get.metaImport{Prefix:"gopkg.in/fsnotify.v1", VCS:"git", RepoRoot:"https://gopkg.in/fsnotify.v1"} at https://gopkg.in/fsnotif
y.v1?go-get=1^@go: finding github.com/nbutton23/zxcvbn-go latest^@go get: error loading module requirements^@```

--- a/autoload/go/util.vim
+++ b/autoload/go/util.vim
@@ -519,6 +519,37 @@ function! go#util#ShowInfo(info)
   echo "vim-go: " | echohl Function | echon a:info | echohl None
 endfunction
 
+" go#util#SetEnv takes the name of an environment variable and what its value
+" should be and returns a function that will restore it to its original value.
+function! go#util#SetEnv(name, value) abort
+  let l:state = {}
+
+  if len(a:name) == 0
+    return function('s:noop', [], l:state)
+  endif
+
+  let l:remove = 0
+  if exists('$' . a:name)
+    let l:oldvalue = eval('$' . a:name)
+  else
+    let l:remove = 1
+  endif
+
+  call execute('let $' . a:name . ' = "' . a:value . '"')
+
+  if l:remove
+    function! s:remove(name) abort
+      call execute('unlet $' . a:name)
+    endfunction
+    return function('s:remove', [a:name], l:state)
+  endif
+
+  return function('go#util#SetEnv', [a:name, l:oldvalue], l:state)
+endfunction
+
+function! s:noop(...) abort dict
+endfunction
+
 " restore Vi compatibility settings
 let &cpo = s:cpo_save
 unlet s:cpo_save

--- a/plugin/go.vim
+++ b/plugin/go.vim
@@ -101,11 +101,11 @@ function! s:GoInstallBinaries(updateBinaries, ...)
   " change $GOBIN so go get can automatically install to it
   let $GOBIN = go_bin_path
 
-  " old_path is used to restore users own path
-  let old_path = $PATH
-
   " vim's executable path is looking in PATH so add our go_bin path to it
-  let $PATH = go_bin_path . go#util#PathListSep() . $PATH
+  let Restore_path = go#util#SetEnv('PATH', go_bin_path . go#util#PathListSep() . $PATH)
+
+  " GO111MODULE must be off to install golanci-lint and gometalinter
+  let Restore_modules = go#util#SetEnv('GO111MODULE', 'off')
 
   " when shellslash is set on MS-* systems, shellescape puts single quotes
   " around the output string. cmd on Windows does not handle single quotes
@@ -185,7 +185,9 @@ function! s:GoInstallBinaries(updateBinaries, ...)
   endfor
 
   " restore back!
-  let $PATH = old_path
+  call call(Restore_path, [])
+  call call(Restore_modules, [])
+
   if resetshellslash
     set shellslash
   endif

